packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs

```diff
@@ -0,0 +1,74 @@
+use std::{fmt, fmt::Formatter, marker::PhantomData};
+
+use crate::ontology::types::{uri::BaseUri, DataType, EntityType, LinkType, PropertyType};
+
+pub type DataTypeQuery<'q> = OntologyQuery<'q, DataType>;
+pub type PropertyTypeQuery<'q> = OntologyQuery<'q, PropertyType>;
+pub type LinkTypeQuery<'q> = OntologyQuery<'q, LinkType>;
+pub type EntityTypeQuery<'q> = OntologyQuery<'q, EntityType>;
+
+#[derive(Debug, Copy, Clone)]
+pub enum OntologyVersion {
+    Exact(u32),
+    Latest,
+}
+
+pub struct OntologyQuery<'q, T> {
+    _marker: PhantomData<fn() -> T>,
+    uri: Option<&'q BaseUri>,
+    version: Option<OntologyVersion>,
+}
+
+impl<T> fmt::Debug for OntologyQuery<'_, T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("OntologyQuery")
+            .field("uri", &self.uri())
+            .field("version", &self.version())
+            .finish()
+    }
+}
+
+impl<T> OntologyQuery<'_, T> {
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            _marker: PhantomData,
+            uri: None,
+            version: None,
+        }
+    }
+}
+
+/// Methods for building up a query.
+impl<'q, T> OntologyQuery<'q, T> {
+    #[must_use]
+    pub const fn with_uri(mut self, uri: &'q BaseUri) -> Self {
+        self.uri = Some(uri);
+        self
+    }
+
+    #[must_use]
+    pub const fn with_version(mut self, version: u32) -> Self {
+        self.version = Some(OntologyVersion::Exact(version));
+        self
+    }
+
+    #[must_use]
+    pub const fn with_latest_version(mut self) -> Self {
```

Comment:
Should this `with_latest_version` (which i guess will become `by_latest_version`) not be different than the others? Instead of being a filter, this is more like an ordering + limiting clause. Perhaps `pick_latest_version` or something would be more appropriate.

Reply:
I wonder if we want to explore this a bit more (genuinely maybe not wort hit right now), I could see us wanting other kinds of queries where we're saying "pick the first/last element when sorted by X". I think this is still a sort of filter, but as you're saying I think it's more a combination of two operations, ones that we don't necessarily have good plans to support yet, or know how to handle them.